centrallix-sysdoc/OSDriver_Authoring.md

```diff
@@ -115,497 +170,1032 @@ License:  Copyright (C) 2001-2011 LightSys Technology Services. See LICENSE.txt
+| [GetFirstAttr](#function-getfirstattr--getnextattr)*      | Get the name of the object's first attribute.
+| [GetNextAttr](#function-getfirstattr--getnextattr)*       | Get the name of the object's next attribute.
+| [SetAttrValue](#function-setattrvalue)                    | Set the value of an object's attribute.
+| [AddAttr](#function-addattr)                              | Add a new attribute to an object.
+| [OpenAttr](#function-openattr)                            | Open an attribute as if it were an object with content.
+| [GetFirstMethod](#function-getfirstmethod--getnextmethod) | Get the name of an object's first method.
+| [GetNextMethod](#function-getfirstmethod--getnextmethod)  | Get the name of an object's next method.
+| [ExecuteMethod](#function-executemethod)                  | Execute a method with a given name and optional parameter string.
+| [PresentationHints](#function-presentationhints)          | Get info about an object's attributes.
+| [Info](#function-info)*                                   | Get info about an object instance.
+| [Commit](#function-commit)                                | Commit changes made to an object.
+| [GetQueryCoverageMask](#function-getquerycoveragemask)    | ???
+| [GetQueryIdentityPath](#function-getqueryidentitypath)    | ???
+
+_*Function is always required._
+
+_**Function is required to support queries._
+
+
+---
+### Abbreviative Prefix
+Each OS Driver will have an abbreviation prefix, such as `qy` for the query driver or `sydb` for the sybase database driver.  This prefix should be prepended to the start of every public function name within the OS driver for consistency and scope management (e.g. `qyInitialize()`, `sydbQueryFetch()`, etc.). Normally, a driver's abbreviation prefix is two to four characters, all lowercase and may be the same as a file extension the driver supports. However, this is not an absolute requirement (see the cluster driver in `objdrv_cluster.c` which supports `.cluster` files using an abbreviation prefix of `cluster`).
+
+This document uses `xxx` to refer to an unspecified abbreviative prefix.
+
+---
+### Internal Functions
+It is highly likely that driver authors will find shared functionality in the following functions, or wish to abstract out functionality from any of them for a variety of reasons.  When creating additional internal functions in this way, they should be named using the convention of `xxx_internal_FunctionName()`, or possibly `xxxi_FunctionName()` for short.
+
+---
+### Function: Initialize
+```c
+/*** @returns 0 if successful, or
+ ***         -1 if an error occurred.
+ ***/
+int xxxInitialize(void)
+```
+- ⚠️ **Warning**: Currently, the success/failure of this function is ignored by the caller.
+- 📖 **Note**: Unlike other functions defined in the driver, each  driver author must manually add this call to the start up code, found in the `cxDriverInit()` function in `centrallix.c`.
```

Comment:
It might be useful to include a discussion of how to create a dynamically loaded module.  Examples include the MySQL and Sybase drivers, both of which are dependent on libraries that may or may not be installed on the target system, and so have to be distributed as separate packages.  This involves some autoconf work too.

Reply:
Agreed, but I don't know how to do that.

This task has been added to my long term todo list.